components/ledger/pkg/ledger/query/worker.go

```diff
@@ -184,125 +291,204 @@ func (w *Worker) initLedger(ctx context.Context, ledger string) error {
 	return nil
 }
 
-func (w *Worker) processLogs(ctx context.Context, store storage.LedgerStore, logs ...core.Log) error {
-	for _, log := range logs {
-		if err := w.processLog(ctx, store, &log); err != nil {
-			return errors.Wrapf(err, "processing log %d", log.ID)
-		}
+func (w *Worker) processLogs(ctx context.Context, logs ...core.Log) error {
+
+	logsData, err := w.buildData(ctx, logs...)
+	if err != nil {
+		return errors.Wrap(err, "building data")
 	}
 
-	return nil
-}
+	if err := w.store.RunInTransaction(ctx, func(ctx context.Context, tx storage.LedgerStore) error {
+		if len(logsData.accountsToUpdate) > 0 {
+			if err := tx.UpdateAccountsMetadata(ctx, logsData.accountsToUpdate); err != nil {
+				return errors.Wrap(err, "updating accounts metadata")
+			}
+		}
 
-func (w *Worker) processLog(ctx context.Context, store storage.LedgerStore, log *core.Log) error {
-	volumeAggregator := aggregator.Volumes(store)
+		if len(logsData.transactionsToInsert) > 0 {
+			if err := tx.InsertTransactions(ctx, logsData.transactionsToInsert...); err != nil {
+				return errors.Wrap(err, "inserting transactions")
+			}
+		}
 
-	var err error
-	switch log.Type {
-	case core.NewTransactionLogType:
-		payload := log.Data.(core.NewTransactionLogPayload)
-		txVolumeAggregator, err := volumeAggregator.NextTxWithPostings(ctx, payload.Transaction.Postings...)
-		if err != nil {
-			return err
+		if len(logsData.transactionsToUpdate) > 0 {
+			if err := tx.UpdateTransactionsMetadata(ctx, logsData.transactionsToUpdate...); err != nil {
+				return errors.Wrap(err, "updating transactions")
+			}
 		}
 
-		if payload.AccountMetadata != nil {
-			for account, metadata := range payload.AccountMetadata {
-				if err := store.UpdateAccountMetadata(ctx, account, metadata); err != nil {
-					return errors.Wrap(err, "updating account metadata")
-				}
+		if len(logsData.ensureAccountsExist) > 0 {
+			if err := tx.EnsureAccountsExist(ctx, logsData.ensureAccountsExist); err != nil {
+				return errors.Wrap(err, "ensuring accounts exist")
 			}
 		}
 
-		expandedTx := core.ExpandedTransaction{
-			Transaction:       payload.Transaction,
-			PreCommitVolumes:  txVolumeAggregator.PreCommitVolumes,
-			PostCommitVolumes: txVolumeAggregator.PostCommitVolumes,
+		if len(logsData.volumesToUpdate) > 0 {
+			return tx.UpdateVolumes(ctx, logsData.volumesToUpdate...)
 		}
 
-		if err := store.InsertTransactions(ctx, expandedTx); err != nil {
-			return errors.Wrap(err, "inserting transactions")
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if w.monitor != nil {
+		for _, monitor := range logsData.monitors {
+			monitor(ctx, w.monitor)
```

Comment:
maybe run in a routine?

Reply:
I'm not a big fan of that since we can have the events not in order ! After discussing IRL, we'll let that like this for now !